### PR TITLE
Fix #61 2D point generation

### LINKED ISSUE #61: Fix 2D point generation
- Include `element_type` from GMSH as an input
- Fix the functions to be compatible for both 2D and 3D which includes:
    - Calculate volume for 2D, find an area using double triangle cosine rules
    - Overload functions to get for 2D or maybe use partial specialization

### PR DESCRIPTION
**Pull request by ezrayst**
- Use partial specialization for 2d and 3d
- Add new input parameters in the JSON file
- Create two regression tests: one for 2d and 3d
- Add volume calculation for 2d, which is just area * 1

### CODE REVIEW

File: src/main.cc [hunk 1]
```diff
@@ -13,18 +13,24 @@ int main(int argc, char** argv) {
...
     std::unique_ptr<IO> io(new IO(argc, argv));
 
-    //! Mesh
-    std::unique_ptr<Mesh<3, 8>> mesh(new GMSH<3, 8>());
-
     //! MaterialProperties
     std::shared_ptr<MaterialProperties> material =
         std::make_shared<MaterialProperties>(io->material_properties());
 
+    // const unsigned Tdim = io->dimension();
```
Comment (ezrayst) [inline src/main.cc]: I can't seem to use this. Can you let me know how to cast from `non-const` to `const`? Thanks!

Comment (kks32) [reply src/main.cc]: You need a `switch` to handle dimensions

Comment (ezrayst) [reply src/main.cc]: Yes but we need to initialize `const` variable and `switch` would not allow cross-initialization. Or maybe I don't get the way you want to do it?

File: include/gmsh.tcc [hunk 2]
```diff
@@ -151,9 +150,14 @@ void GMSH<Tdim, Tvertices>::read_elements(std::ifstream& file) {
       if (elementtype != element_type) {
         istream >> line;
       } else {
-        istream >> elementarray[0] >> elementarray[1] >> elementarray[2] >>
-            elementarray[3] >> elementarray[4] >> elementarray[5] >>
-            elementarray[6] >> elementarray[7];
+        if (Tdim == 2) {
```
Comment (kks32) [inline include/gmsh.tcc]: Can we try to rewrite this based on the size of the `elementarray` and iterate through that?

File: src/element.cc [hunk 3]
```diff
@@ -4,43 +4,82 @@
...
-      (1.0 / 12) *
-          (a - g).dot(((b - d).cross(c - a)) + ((e - b).cross(f - a)) +
-                      ((d - e).cross(h - a))) +
-      (1.0 / 12) *
-          (b - g).dot(((b - d).cross(c - a)) + ((c - g).cross(c - f))) +
-      (1.0 / 12) *
-          (e - g).dot(((e - b).cross(f - a)) + ((f - g).cross(h - f))) +
-      (1.0 / 12) * (d - g).dot(((d - e).cross(h - a)) + ((h - g).cross(h - c)));
+    
+  } else {
```
Comment (kks32) [inline src/element.cc]: Always check for 3D as well

File: tests/gmsh_test_2d.cc [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+  REQUIRE(stresses.at(0)[5] == Approx(0).epsilon(tolerance));
+  REQUIRE(stresses.at(1)[0] == Approx(0).epsilon(tolerance));
+  REQUIRE(stresses.at(1)[1] == Approx(0).epsilon(tolerance));
+  REQUIRE(stresses.at(1)[2] == Approx(0).epsilon(tolerance));
+  REQUIRE(stresses.at(1)[3] == Approx(0).epsilon(tolerance));
+  REQUIRE(stresses.at(1)[4] == Approx(0).epsilon(tolerance));
+  REQUIRE(stresses.at(1)[5] == Approx(0).epsilon(tolerance));
+  
+  //! Check volume
+  REQUIRE(volumes.at(0) == Approx(1).epsilon(tolerance));
```
Comment (kks32) [inline tests/gmsh_test_2d.cc]: Could you check for non-unit volumes in tests, because it's easy to miss some bugs checking for an unit volume.

**Review by kks32 — CHANGES_REQUESTED** [review]
Looks good, some small changes to make.

**Comment by ezrayst** [comment]
How about getting the `Tdim` and `Tvertices` in `main.cc`?
I think I am happier once this is done.

File: src/main.cc [hunk 5]
```diff
@@ -13,18 +13,47 @@ int main(int argc, char** argv) {
...
+    // const unsigned Tvertices = io->nvertices();
+    // unsigned TTdim = io->dimension();
+    // unsigned Tvertices;
+        
+    // int dim = 2;
+    // switch (dim) {
+    //     case 2: {
+    //         const unsigned Tdim = 2;
```
Comment (kks32) [inline src/main.cc]: The rest of the main should go here and repeat the main for case 3 as well

Comment (ezrayst) [reply src/main.cc]: Understood. For now just leave `Tvertices` together? or you want to have `switch` inside `switch`?

Comment (kks32) [reply src/main.cc]: Isn't `Tvertices` fixed for 2D and 3D?

File: bin/hex_cone.json [hunk 6]
```diff
@@ -5,5 +5,8 @@
 	"material_properties" : {
 		"density" : 2000.0,
 		"k0" : 0.5
-	}
+	},
+	"element_type" : 5,
+	"dimension" : 3,
+	"vertices" : 8
```
Comment (kks32) [inline bin/hex_cone.json]: Shouldn't element type define vertices? or am I missing something?

File: include/gmsh.tcc [hunk 7]
```diff
@@ -194,12 +194,111 @@ void GMSH<Tdim, Tvertices>::store_element_vertices() {
...
-void GMSH<Tdim, Tvertices>::compute_material_points(unsigned ngauss_points) {
+//! \param[in] ngauss_points Number of gauss points per coordinate, element_type GMSH element type
+template <>
+inline void GMSH<2, 4>::compute_material_points(unsigned ngauss_points, unsigned element_type) {
+
+  const unsigned Tdim = 2;
+  const unsigned Tvertices = 4;
+
+  //! Storing ngauss_points to member variable and get constants from namespace
+  ngauss_points_ = ngauss_points;
```
Comment (kks32) [inline include/gmsh.tcc]: Why do we need this member variable `ngauss_points_`?

File: include/gmsh.tcc [hunk 8]
```diff
@@ -194,12 +194,111 @@ void GMSH<Tdim, Tvertices>::store_element_vertices() {
...
+  std::vector<double> gauss_constants =
+      element::gauss_points.find(ngauss_points_)->second;
+
+  //! Create a matrix of xi from gauss points
+  //! Matrix is size npoints x Tdim
+  //! For 2D only
+  unsigned npoints = std::pow(ngauss_points_, Tdim);
+  Eigen::MatrixXd xi_gauss_points(npoints, Tdim);
+  unsigned counter = 0;
+  for (unsigned ii = 0; ii < ngauss_points_; ++ii) {
```
Comment (kks32) [inline include/gmsh.tcc]: Use iteration variables as `i` and `j`

File: include/gmsh.tcc [hunk 9]
```diff
@@ -194,12 +194,111 @@ void GMSH<Tdim, Tvertices>::store_element_vertices() {
...
+  //! global_id is the index for all points
+  unsigned global_id = 0;
+  //! material_id is the index of materialpoints
+  unsigned material_id = 0;
+  //! point_id is local index of points
+  unsigned point_id = 0;
+
+  //! Update vector of material points
+  //! Fill materialpoints_ vector for the first component
+  materialpoints_.emplace_back(std::shared_ptr<MaterialPoints<Tdim>>(
```
Comment (kks32) [inline include/gmsh.tcc]: Can we use `unique_ptr` and use C++14 make_unique

File: include/gmsh.tcc [hunk 10]
```diff
@@ -194,12 +194,111 @@ void GMSH<Tdim, Tvertices>::store_element_vertices() {
       << "The coordinates for vertices of each element have been stored\n";
 }
 
-//! Compute material points based on the centroid
+//! Compute material points for 2D quadrilateral
```
Comment (kks32) [inline include/gmsh.tcc]: The function name is not clear, what is it supposed to do?

Comment (kks32) [reply include/gmsh.tcc]: What do we mean by `compute`? what are we computing?

File: include/gmsh.tcc [hunk 11]
```diff
@@ -194,12 +194,111 @@ void GMSH<Tdim, Tvertices>::store_element_vertices() {
...
+  //! Update vector of material points
+  //! Fill materialpoints_ vector for the first component
+  materialpoints_.emplace_back(std::shared_ptr<MaterialPoints<Tdim>>(
+      new MaterialPoints<Tdim>(material_id)));
+  for (const auto& element : elements_) {
+
+    //! Store coordinates in Tdim x Tvertices matrix
+    //! Where N is the number of nodes per element
+    //! This is rearranging of the data to have stored in matrix form
+    Eigen::MatrixXd node_coordinates(Tdim, Tvertices);
```
Comment (kks32) [inline include/gmsh.tcc]: Do we need to store this matrix? Can we not call `vertex co-ordinates` directly in #269?

File: include/gmsh.tcc [hunk 12]
```diff
@@ -194,12 +194,111 @@ void GMSH<Tdim, Tvertices>::store_element_vertices() {
...
+
+      //! Update point_id
+      ++point_id;
+      //! Update global_id
+      global_id = point_id;
+    }
+  }
+
+  //! Find number of material points generated
+  for (const auto& materialpoint : materialpoints_)
```
Comment (kks32) [inline include/gmsh.tcc]: This naming is unclear, what is `materialpoint`?

File: include/io.h [hunk 13]
```diff
@@ -28,6 +28,15 @@ class IO {
   //! Return the number of points per coordinate
   unsigned ngauss_points() const { return ngauss_points_; }
 
+  //! Return the element type
```
Comment (kks32) [inline include/io.h]: Why should an `io` class return `element_type`, `nvertices`?

File: src/main.cc [hunk 14]
```diff
@@ -13,30 +13,66 @@ int main(int argc, char** argv) {
     //! IO
     std::unique_ptr<IO> io(new IO(argc, argv));
 
-    //! Mesh
-    std::unique_ptr<Mesh<3, 8>> mesh(new GMSH<3, 8>());
-
     //! MaterialProperties
     std::shared_ptr<MaterialProperties> material =
```
Comment (kks32) [inline src/main.cc]: Always use `unique_ptr` unless it will have a shared ownership

File: src/main.cc [hunk 15]
```diff
@@ -13,30 +13,66 @@ int main(int argc, char** argv) {
...
+            mesh->write_volumes(io->output_file("volumes", ".txt"));
+        
+            //! Write .vtk output files for viewing
+            mesh->write_vtk_stresses(io->output_file("initial_stresses", ".vtk"));
+            mesh->write_vtk_mesh(io->output_file("mesh", ".vtk"));
 
+            break;
+        }
+    }
+
```
Comment (kks32) [inline src/main.cc]: Needs a `default` case

**Review by kks32 — CHANGES_REQUESTED** [review]

File: bin/hex_cone.json (hunk 6 above)
Comment (ezrayst) [reply bin/hex_cone.json]: Nope, we can have higher order elements too. For instance, 4-node quadrilateral is a `3`, 9-node quadrilateral is `10` and 8-node quadrilateral is `16`. So in a sense you will have more vertices.

More info is here: http://gmsh.info/doc/texinfo/gmsh.html#File-formats

Comment (kks32) [reply bin/hex_cone.json]: Yes, but why should I have to specify that in the input JSON file, can we not deduct the type of element from the mesh file?

File: include/gmsh.tcc (hunk 7 above)
Comment (ezrayst) [reply include/gmsh.tcc]: Not anymore. I can remove it

File: include/gmsh.tcc (hunk 10 above)
Comment (ezrayst) [reply include/gmsh.tcc]: I would use `generate` then

File: include/gmsh.tcc (hunk 12 above)
Comment (ezrayst) [reply include/gmsh.tcc]: isn't this quite typical for range-for loop?

File: include/io.h (hunk 13 above)
Comment (ezrayst) [reply include/io.h]: Because the element type as previously discussed does make different vertices (nodes)

File: src/main.cc (hunk 14 above)
Comment (ezrayst) [reply src/main.cc]: material properties are shared within material points. So has to be shared

File: include/gmsh.tcc (hunk 11 above)
Comment (ezrayst) [reply include/gmsh.tcc]: I think we can, but to me, the loop after that would not be as elegant. I feel that we make it rather cleaner using matrix

File: bin/hex_cone.json (hunk 6 above)
Comment (ezrayst) [reply bin/hex_cone.json]: Understood your concern now. Hmm, my concern is we have templatized `Tvertices` and we need them. Should we do something in `io` to store `nvertices` given certain `element_type`? I feel that is quite a hassle too

File: include/gmsh.tcc (hunk 9 above)
Comment (ezrayst) [reply include/gmsh.tcc]: Since we use C++11, do you want to make a function that kinds of use C++14 make_unique?

File: bin/hex_cone.json (hunk 6 above)
Comment (kks32) [reply bin/hex_cone.json]: `IO` class is `IO` it shouldn't worry about `nvertices`. Use a factory to create the right `element_type`

File: include/gmsh.tcc (hunk 9 above)
Comment (kks32) [reply include/gmsh.tcc]: No, we will just use C++14

File: include/gmsh.tcc (hunk 11 above)
Comment (kks32) [reply include/gmsh.tcc]: OK

File: include/gmsh.tcc (hunk 12 above)
Comment (kks32) [reply include/gmsh.tcc]: `materialpoints_` is a collection of material points, so I can see why it has `npoints`, but I don't understand why a single `materialpoint` has `npoints`

File: include/io.h (hunk 13 above)
Comment (kks32) [reply include/io.h]: `IO` class shouldn't worry about `element_type`, `nvertices` etc. They are the problem of the mesh. My issue is with abstraction, who owns what. Why should IO have anything to do with what `element_type` the input has, it should be the problem of the mesh.

File: src/main.cc (hunk 14 above)
Comment (kks32) [reply src/main.cc]: I find shared ownership of material properties difficult to understand. Feels not correct. Mesh owns material points, where the ownership is clear. Material points don't own material models. That's the problem I'm having. What's the best way? Maybe a const pointer to material model.

**Review by kks32 — CHANGES_REQUESTED** [review]

File: include/gmsh.h [hunk 16]
```diff
@@ -34,21 +34,18 @@ class GMSH : public Mesh<Tdim, Tvertices> {
...
 
  private:
   //! Total number of vertices
   using Mesh<Tdim, Tvertices>::nvertices_;
 
-  //! Number of gauss points per coordinate
-  using Mesh<Tdim, Tvertices>::ngauss_points_;
-
   //! Total number of material points generated
   using Mesh<Tdim, Tvertices>::npoints_;
```
Comment (kks32) [inline include/gmsh.h:50]: What's the difference between `npoints_` and `materialpoints_`

File: include/gmsh.tcc [hunk 17]
```diff
@@ -194,29 +197,130 @@ void GMSH<Tdim, Tvertices>::store_element_vertices() {
...
+//! GMSH element type
+template <>
+inline void GMSH<2, 4>::generate_material_points(unsigned ngauss_points,
+                                                 unsigned element_type) {
+
+  const unsigned Tdim = 2;
+  const unsigned Tvertices = 4;
+
+  //! Get constants from namespace
+  std::vector<double> gauss_constants =
```
Comment (kks32) [inline include/gmsh.tcc]: `auto`

File: include/gmsh.tcc [hunk 18]
```diff
@@ -194,29 +197,130 @@ void GMSH<Tdim, Tvertices>::store_element_vertices() {
...
   Eigen::MatrixXd xi_gauss_points(npoints, Tdim);
   unsigned counter = 0;
-  for (unsigned ii = 0; ii < ngauss_points_; ++ii) {
-    for (unsigned jj = 0; jj < ngauss_points_; ++jj) {
-      for (unsigned kk = 0; kk < ngauss_points_; ++kk) {
-        xi_gauss_points(counter, 0) = gauss_constants.at(ii);
-        xi_gauss_points(counter, 1) = gauss_constants.at(jj);
+  for (unsigned i = 0; i < ngauss_points; ++i) {
+    for (unsigned j = 0; j < ngauss_points; ++j) {
+      for (unsigned kk = 0; kk < ngauss_points; ++kk) {
```
Comment (kks32) [inline include/gmsh.tcc]: change to `k`

File: include/gmsh.tcc [hunk 19]
```diff
@@ -194,29 +197,130 @@ void GMSH<Tdim, Tvertices>::store_element_vertices() {
...
+//! GMSH element type
+template <>
+inline void GMSH<3, 8>::generate_material_points(unsigned ngauss_points,
+                                                 unsigned element_type) {
+
+  const unsigned Tdim = 3;
+  const unsigned Tvertices = 8;
+
+  //! Get constants from namespace
+  ngauss_points = ngauss_points;
```
Comment (kks32) [inline include/gmsh.tcc]: what's this line above?

File: include/gmsh.tcc [hunk 20]
```diff
@@ -141,6 +139,7 @@ void GMSH<Tdim, Tvertices>::read_elements(std::ifstream& file,
   //! 5 - Hexahedron (8 nodes)
   //! For more informtion on element types, visit:
   //! http://gmsh.info/doc/texinfo/gmsh.html#File-formats
+  unsigned element_type = 3;
```
Comment (kks32) [inline include/gmsh.tcc]: Isn't this just the 2D case?

File: include/gmsh.tcc [hunk 21]
```diff
@@ -200,17 +199,16 @@ void GMSH<Tdim, Tvertices>::store_element_vertices() {
...
-inline void GMSH<2, 4>::generate_material_points(unsigned ngauss_points,
-                                                 unsigned element_type) {
+inline void GMSH<2, 4>::generate_material_points(unsigned ngauss_points) {
 
   const unsigned Tdim = 2;
   const unsigned Tvertices = 4;
 
   //! Get constants from namespace
-  std::vector<double> gauss_constants =
+  auto gauss_constants =
```
Comment (kks32) [inline include/gmsh.tcc]: `clang-format`

File: include/io.h [hunk 22]
```diff
@@ -63,14 +57,8 @@ class IO {
   //! Number of gauss points per coordinate
   unsigned ngauss_points_{0};
```
Comment (kks32) [inline include/io.h]: We don't need to store this variable

File: tests/gmsh_test_2d.cc [hunk 23]
```diff
@@ -28,8 +28,8 @@ TEST_CASE("GMSH is checked in 2D", "[GMSH][2D]") {
   auto material = std::shared_ptr<MaterialProperties>(
```
Comment (kks32) [inline tests/gmsh_test_2d.cc]: Can we make this unique_ptr?

File: tests/gmsh_test_3d.cc [hunk 24]
```diff
@@ -28,8 +28,8 @@ TEST_CASE("GMSH is checked in 3D", "[GMSH][3D]") {
   auto material = std::shared_ptr<MaterialProperties>(
```
Comment (kks32) [inline tests/gmsh_test_3d.cc]: Can we make this unique_ptr?

**Review by kks32 — CHANGES_REQUESTED** [review]

**Comment by ezrayst** [comment]
@kks32, you might be making comments at the outdated file, Krishna.

**Comment by ezrayst** [comment]
Structure is correct I think, but implementation for more than one material property in the body is still lacking.

File: include/gmsh.tcc [hunk 25]
```diff
@@ -194,30 +202,127 @@ void GMSH<Tdim, Tvertices>::store_element_vertices() {
...
 //! \tparam Tdim Dimension
 //! \tparam Tvertices Number of vertices in element
 //! \param[in] ngauss_points Number of gauss points per coordinate
-template <unsigned Tdim, unsigned Tvertices>
-void GMSH<Tdim, Tvertices>::compute_material_points(unsigned ngauss_points) {
+//! GMSH element type
+template <>
+inline void GMSH<2, 4>::generate_material_points(unsigned ngauss_points) {
+
+  const unsigned Tdim = 2;
```
Comment (kks32) [inline include/gmsh.tcc]: This is not a template parameter, so please use dim and vertices

File: include/gmsh.tcc [hunk 26]
```diff
@@ -194,30 +202,127 @@ void GMSH<Tdim, Tvertices>::store_element_vertices() {
...
+  //! global_id is the index for all points
+  unsigned global_id = 0;
+  //! material_id is the index of materialpoints
+  unsigned material_id = 0;
+  //! point_id is local index of points
+  unsigned point_id = 0;
+
+  //! Update vector of material points
+  //! Fill materialpoints_ vector for the first component
+  materialpoints_.emplace_back(std::unique_ptr<MaterialPoints<Tdim>>(
```
Comment (kks32) [inline include/gmsh.tcc]: `make_unique` and move to `materialpoints_`

File: include/gmsh.tcc (hunk 12 above)
Comment (kks32) [reply include/gmsh.tcc]: Could you iterate from `materialpoints : materialpoints_` it's not a singular material point

File: include/gmsh.tcc [hunk 27]
```diff
@@ -194,30 +202,127 @@ void GMSH<Tdim, Tvertices>::store_element_vertices() {
...
+//! \param[in] ngauss_points Number of gauss points per coordinate
+//! GMSH element type
+template <>
+inline void GMSH<3, 8>::generate_material_points(unsigned ngauss_points) {
+
+  const unsigned Tdim = 3;
+  const unsigned Tvertices = 8;
+
+  //! Get constants from namespace
   std::vector<double> gauss_constants =
```
Comment (kks32) [inline include/gmsh.tcc]: auto

File: include/io.h [hunk 28]
```diff
@@ -26,13 +26,16 @@ class IO {
...
   //! Return the number of points per coordinate
-  unsigned ngauss_points() const { return ngauss_points_; }
+  unsigned ngauss_points() const;
+
+  //! Return the dimension of the problem: 2 or 3
+  unsigned dimension() const;
 
   //! Return json object for material properties
-  json material_properties() const { return json_material_properties_; }
+  std::vector<json> material_properties() const;
```
Comment (kks32) [inline include/io.h]: Why is this a vector of `json` objects?

File: include/io.h [hunk 29]
```diff
@@ -44,15 +47,6 @@ class IO {
 
   //! Input json object
   json json_;
```
Comment (kks32) [inline include/io.h]: type is `Json`

File: include/mesh.h [hunk 30]
```diff
@@ -30,11 +30,11 @@ class Mesh {
   virtual void read_mesh(const std::string& filename) = 0;
 
   //! Compute material point location
-  virtual void compute_material_points(unsigned ngauss_points) = 0;
+  virtual void generate_material_points(unsigned ngauss_points) = 0;
 
   //! Get material properties from json object
   void assign_material_properties(
-      const std::shared_ptr<MaterialProperties>& material);
+      std::vector<std::unique_ptr<MaterialProperties>> material);
```
Comment (kks32) [inline include/mesh.h:37]: If this is receiving material properties from JSON object, this should get a JSON object

File: src/io.cc [hunk 31]
```diff
@@ -53,39 +50,84 @@ IO::IO(int argc, char** argv) {
...
   meshfile.close();
 
-  //! Read and store number of gauss points per coordinate
+  return mesh_file_name;
+}
+
+//! \brief Return user-specified material propertes
+std::vector<json> IO::material_properties() const {
+
+  std::vector<json> json_material_properties;
```
Comment (kks32) [inline src/io.cc]: Why not just return a single JSON object of material properties?

File: src/main.cc [hunk 32]
```diff
@@ -13,29 +14,97 @@ int main(int argc, char** argv) {
...
-    //! Mesh
-    std::unique_ptr<Mesh<3, 8>> mesh(new GMSH<3, 8>());
-
     //! MaterialProperties
-    std::shared_ptr<MaterialProperties> material =
-        std::make_shared<MaterialProperties>(io->material_properties());
+    std::vector<std::unique_ptr<MaterialProperties>> material;
+
+    for (unsigned i = 0; i < io->material_properties().size(); i++) {
+      material.emplace_back(
```
Comment (kks32) [inline src/main.cc:21]: Who owns the material?

File: src/main.cc [hunk 33]
```diff
@@ -13,29 +14,97 @@ int main(int argc, char** argv) {
...
+
+        //! Write .vtk output files for viewing
+        mesh->write_vtk_stresses(io->output_file("initial_stresses", ".vtk"));
+        mesh->write_vtk_mesh(io->output_file("mesh", ".vtk"));
+
+        break;
+      }
+
+      default: {
+        const unsigned Tdim = 2;
```
Comment (kks32) [inline src/main.cc]: `default` should just be break;

Comment (kks32) [reply src/main.cc]: or throw an exception

**Review by kks32 — CHANGES_REQUESTED** [review]

File: src/io.cc (hunk 31 above)
Comment (ezrayst) [reply src/io.cc]: We want to know the number of `material properties` right?
It's not that we have `density` and `k0` as a vector, but they together become one.

File: include/gmsh.tcc (hunk 26 above)
Comment (ezrayst) [reply include/gmsh.tcc]: I think it would be better to have `make_unique` in `emplace_back`?

File: src/main.cc (hunk 32 above)
Comment (ezrayst) [reply src/main.cc:21]: At this stage, `main` owns it but it is later moved to `mesh` and then `materialpoints`.
So in the end only `materialpoints` owns the `points` with its coordinates and ids, and the `materialproperties` too

File: src/io.cc (hunk 31 above)
Comment (ezrayst) [reply src/io.cc]: Any issue (performance? memory) with inputting and outputting vector?

File: include/mesh.h (hunk 30 above)
Comment (ezrayst) [reply include/mesh.h:37]: We can call constructor of `MaterialProperties` with input of the JSON object

Comment (ezrayst) [reply include/mesh.h:37]: Another way is we can refactor them and just output and input JSON object, but make `MaterialProperties` in `mesh`. Would that be cleaner?

**Review by kks32 — APPROVED** [review]

**Pull request merged**